src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
-          if(element_size<=sub_size)
-          {
-            new_value = src.value();
-            stored_value_offset=zero_offset;
-          }
-          else
-          {
-            // Offset to retrieve from the stored value: make sure not to
-            // ask for anything out of range; in the first- or last-cell cases
-            // we will adjust the offset in the byte_update call below.
-            if(is_first_cell)
-              stored_value_offset=zero_offset;
-            else if(is_last_cell)
-              stored_value_offset=
-                from_integer(element_size-sub_size, offset_type);
-            else
-              stored_value_offset = minus_exprt(cell_offset, src.offset());
-
-            INVARIANT(
-              src.id() == ID_byte_update_little_endian ||
-                src.id() == ID_byte_update_big_endian,
-              "byte update expression should either be little or big endian");
-
-            byte_extract_exprt byte_extract_expr(
-              src.id() == ID_byte_update_little_endian
-                ? ID_byte_extract_little_endian
-                : ID_byte_extract_big_endian,
-                src.value(), stored_value_offset,
-              element_size < sub_size ? src.value().type() : subtype);
-
-            new_value=lower_byte_extract(byte_extract_expr, ns);
-          }
-
-          // Where does the value we just extracted align in this cell?
-          // This value might be negative, indicating only the most-significant
-          // bytes should be written, to the least-significant bytes of the
-          // target array cell.
-          exprt overwrite_offset;
-          if(is_first_cell)
-            overwrite_offset = mod_exprt(src.offset(), sub_size_expr);
-          else if(is_last_cell)
-          {
-            // This is intentionally negated; passing is_last_cell
-            // to flatten below leads to it being negated again later.
-            overwrite_offset = minus_exprt(
-              minus_exprt(cell_offset, src.offset()), stored_value_offset);
-          }
-          else
-            overwrite_offset=zero_offset;
-
-          byte_update_exprt byte_update_expr(
-            src.id(),
-            old_cell_value,
-            overwrite_offset,
-            new_value);
-
-          // Call recursively, the array is gone!
-          // The last parameter indicates that the
-          exprt flattened_byte_update_expr=
-            lower_byte_update(byte_update_expr, ns, is_last_cell);
-
-          with_exprt with_expr(
-            result, cell_index, flattened_byte_update_expr);
-
-          result=with_expr;
-        }
-
-        return simplify_expr(result, ns);
-      }
+  if(is_array)
+    return simplify_expr(
+      array_exprt(std::move(elements), to_array_type(src.type())), ns);
+  else
+    return simplify_expr(
+      vector_exprt(std::move(elements), to_vector_type(src.type())), ns);
+}
+
+static exprt lower_byte_update_struct(
```

Comment:
new function should be documented

Reply:
Done (all newly added functions are now documented).